File: src/etch/bindings/csharp/transport/Listener.cs
using System;
using System.Text;
using System.Net;
using System.Net.Sockets;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using Etch.Support;
using Etch.Util;

namespace Etch.Transport
{
    /// <summary>
    /// Implementation of a connection which handles a socket listener.
    /// </summary>
    public class Listener : Connection<SessionListener>,Transport<SessionListener>
    {

        /// <summary>
        /// Constructs the Listener.
        /// </summary>
        /// <param name="backlog">max number of pending connection requests.</param>
        /// <param name="host">address to listen to. Null means any local address.</param>
        /// <param name="port">port to listen on. Port must be >= 0. Port of 0
        /// means allocate an available port.</param>
        /// <param name="delay">how long in milliseconds to wait before retrying a
        /// failure. Delay must be >= 0. Delay of 0 means do not retry.</param>

        public Listener( int backlog, String host,
            int port, int delay ) 
        {
            if ( backlog < 0 )
                throw new ArgumentException( "backlog < 0" );

            if ( port < 0 || port > 65535 )
                throw new ArgumentException( "port < 0 || port > 65535" );

            if ( delay < 0 )
                throw new ArgumentException( "delay < 0" );

            this.backlog = backlog;
            this.host = host;
            this.port = port;
            this.delay = delay;
        }

        /// <summary>
        /// Query term from URI to specify backlog value to ServerSocket. The
        /// value is "TcpListener.backlog".
        /// </summary>
        public readonly static String BACKLOG = "TcpListener.backlog";

       
        private Listener( URL uri ) : this( (int)uri.GetIntegerTerm( BACKLOG, 0 ), TranslateHost( uri.Host ), uri.Port, 0 )
        {}

        /// <summary>
        /// Constructs a Listener, initialized from the URI. Listens on the host
        /// and port specified in the URI. To listen on all interfaces, specify
        /// host as ALL_INTFS ("0.0.0.0"). If port is specified or defaulted to 0,
        /// an unused port will be selected.
        /// </summary>
        /// <param name="uri"></param>

        public Listener (string uri) : this (new URL(uri))
        {
            
        }

        private int backlog;
	
	    private String host;
    	
	    private int port;
    	
	    private int delay;

        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        
        protected override void Stop0()
        {
            Close( true );
            base.Stop0();
        }
        
        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        /// <returns></returns>
        private Socket CheckSocket()
        {
            Socket ss = serverSocket;
            if ( ss == null )
            {
                Console.WriteLine( "----socket was null----" );
                throw new SocketException();
            }
            return ss;
        }

        private Socket serverSocket;

        public override string ToString()
        {
            if ( serverSocket == null )
                return String.Format("Listener (down, {0}, {1})", host, port);

            return String.Format("Listener (up, {0})", serverSocket.LocalEndPoint);
        }

        [MethodImpl (MethodImplOptions.Synchronized)]
        protected override bool OpenSocket( bool reconnect )
        {
            bool first = true;
            while ( IsStarted() )
            {
                if ( reconnect || !first )
                {
                    if ( delay == 0 )
                        return false;

                    System.Threading.Monitor.Wait(this, delay);

                    if ( !IsStarted() )
                        break;
                }

                try
                {
                    //IPHostEntry ipEntry = Dns.GetHostEntry( host );

                    IPAddress addr;
                    if ( host != null )
                    {
                        IPAddress[] addrs = Dns.GetHostAddresses( host );
                        if ( addrs == null || addrs.Length == 0 )
                            throw new ArgumentException( "host is invalid" );
                        addr = addrs[ 0 ];
                    }
                    else 
                    {
                        addr = IPAddress.Any;
                    }
                    
                    //IPAddress[] addr = ipEntry.AddressList;
                    //string sIPAddress = addr[ 0 ].ToString();

                    IPEndPoint ipe = new IPEndPoint( addr, port );
                    
                    serverSocket = new Socket( ipe.AddressFamily, SocketType.Stream, ProtocolType.Tcp );

                    serverSocket.Bind( ipe );
                    serverSocket.Listen( backlog );

                    return true;
                }
                //catch ( Exception e )
                catch (SocketException e )
                {
                    if ( first )
                    {
                        first = false;
                        FireException( "open", e );
                    }
                }
                
            }
            return false;
        }

        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        protected override void SetUpSocket()
        {
            // nothing to do.
        }

        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        protected override void ReadSocket()
        {
            Socket ss = CheckSocket();
            while ( IsStarted() )
            {
                Socket s = ss.Accept();

                try
                {
                   
                    session.SessionAccepted(s);
                }
                catch ( Exception e )
                {
                    s.Close();
                    FireException("accepted", e);
                }
            }
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="reset"></param>
        /// Exception:
        ///     throws Exception
        public override void Close( bool reset )
        {
            Socket ss = serverSocket;
            if ( ss != null )
            {
                serverSocket = null;
                ss.Close();
            }
        }

        public override EndPoint LocalAddress()
        {
            return CheckSocket().LocalEndPoint;
        }

        public override EndPoint RemoteAddress()
        {
            // ignore
            return null;
        }

        public void Data( Who recepient, FlexBuffer buf )
        {
            // ignore
        }


    /*    public void SetHandler( DataHandler handler )
        {
            base.SetHandler( handler );
            handler.SetSource( this );
        } */

       

       
       
    }
}

File: src/metreos/core/io/Listener.java
/* $Id$
 *
 * Created by wert on Mar 26, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package metreos.core.io;

import java.io.IOException;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.SocketException;

import metreos.util.URL;

/**
 * Implementation of a connection which handles a socket listener.
 */
public class Listener extends Connection<ListenerHandler>
{
	/**
	 * Constructs the Listener.
	 * @param handler target to receive listener events.
	 * @param backlog max number of pending connection requests.
	 * @param host address to listen to. Null means any local address.
	 * @param port port to listen on. Port must be >= 0. Port of 0
	 * means allocate an available port.
	 * @param delay how long in milliseconds to wait before retrying a
	 * failure. Delay must be >= 0. Delay of 0 means do not retry.
	 */
	public Listener( ListenerHandler handler, int backlog, String host,
		int port, int delay )
	{
		super( handler );
		
		if (backlog < 0)
			throw new IllegalArgumentException( "backlog < 0" );
		
		if (port < 0 || port > 65535)
			throw new IllegalArgumentException( "port < 0 || port > 65535" );
		
		if (delay < 0)
			throw new IllegalArgumentException( "delay < 0" );
		
		this.backlog = backlog;
		this.host = host;
		this.port = port;
		this.delay = delay;
	}
	
	/**
	 * Query term from URI to specify backlog value to ServerSocket. The
	 * value is "TcpListener.backlog".
	 * @see #Listener(URL)
	 */
	public final static String BACKLOG = "TcpListener.backlog";
	
	/**
	 * Constructs a Listener, initialized from the URI. Listens on the host
	 * and port specified in the URI. To listen on all interfaces, specify
	 * host as ALL_INTFS ("0.0.0.0"). If port is specified or defaulted to 0,
	 * an unused port will be selected.
	 * @param uri
	 * @see URL#getHost()
	 * @see URL#getPort()
	 * @see URL#getIntegerTerm(String, int)
	 * @see #BACKLOG
	 * @see #ALL_INTFS
	 * @see #localAddress()
	 */
	public Listener( URL uri )
	{
		this( null, uri.getIntegerTerm( BACKLOG, 0 ),
			translateHost( uri.getHost() ), uri.getPort(), 0 );
	}

	private final int backlog;
	
	private final String host;
	
	private final int port;
	
	private final int delay;
	
	@Override
	public void setHandler( ListenerHandler handler )
	{
		super.setHandler( handler );
		handler.setSource( this );
	}
	
	@Override
	protected void stop0() throws Exception
	{
		close( true );
		super.stop0();
	}
	
	private ServerSocket checkSocket() throws IOException
	{
		ServerSocket ss = serverSocket;
		if (ss == null)
			throw new SocketException( "socket closed" );
		return ss;
	}

	private ServerSocket serverSocket;
	
	@Override
	public String toString()
	{
		if (serverSocket == null)
			return String.format( "Listener(down, %s, %d)", host, port );
		
		return String.format( "Listener(up, %s, %d)",
			serverSocket.getInetAddress(), serverSocket.getLocalPort() );
	}

	@Override
	protected synchronized boolean openSocket( boolean reconnect ) throws Exception
	{
		boolean first = true;
		while (isStarted())
		{
			if (reconnect || !first)
			{
				if (delay == 0)
					return false;
				
				wait( delay );
				
				if (!isStarted())
					break;
			}
			
			try
			{
				serverSocket = new ServerSocket( port, backlog,
					host != null ? InetAddress.getByName( host ) : null );
				return true;
			}
			catch ( Exception e )
			{
				if (first)
				{
					first = false;
					fireException( "open", e );
				}
			}
		}
		return false;
	}

	@Override
	protected void setupSocket() throws Exception
	{
		// nothing to do.
	}

	@Override
	protected void readSocket() throws Exception
	{
		ServerSocket ss = checkSocket();
		while (isStarted())
		{
			Socket s = ss.accept();
			try
			{
				handler.accepted( s );
			}
			catch ( Exception e )
			{
				s.close();
				fireException( "accepted", e );
			}
		}
	}

	@Override
	public void close( boolean reset ) throws Exception
	{
		ServerSocket ss = serverSocket;
		if (ss != null)
		{
			serverSocket = null;
			ss.close();
		}
	}

	public SocketAddress localAddress() throws IOException
	{
		return checkSocket().getLocalSocketAddress();
	}

	public SocketAddress remoteAddress()
	{
		// ignore.
		return null;
	}
}
